dsls/sql/src/main/java/org/apache/beam/dsls/sql/BeamSql.java

```diff
@@ -0,0 +1,292 @@
+ * Beam SQL CLI client(pending) is another way to leverage SQL to process data,
+ * which is built on DSL interfaces.
+ *
+ */
+public class BeamSql implements Serializable {
+  private static final Logger LOG = LoggerFactory.getLogger(BeamSql.class);
+
+  private static SchemaPlus schema;
+  private static BeamQueryPlanner planner;
+
+  static{
+    //Disable Assertion in Calcite temporally, it would not ignore any grammar errors.
+    ClassLoader.getSystemClassLoader().setDefaultAssertionStatus(false);
+    schema = Frameworks.createRootSchema(true);
+    planner = new BeamQueryPlanner(schema);
+  }
+
+  /**
+   * Add a UDF function.
+   */
+  public static void addUDFFunction(String functionName, Class<?> className, String methodName){
+    schema.add(functionName, ScalarFunctionImpl.create(className, methodName));
+  }
+
+  /**
+   * add a {@link BaseBeamTable} to schema repository.
+   *
+   */
+  public static void registerTable(String tableName, BaseBeamTable table) {
+    schema.add(tableName, table);
+    planner.getSourceTables().put(tableName, table);
+  }
+
+  /**
+   * Find {@link BaseBeamTable} by table name.
+   */
+  public static BaseBeamTable findTable(String tableName){
+    return planner.getSourceTables().get(tableName);
+  }
+  /**
+   * explain and display the logical execution plan.
+   */
+  public static String explainQuery(String sqlString)
+      throws ValidationException, RelConversionException, SqlParseException {
+    BeamRelNode exeTree = planner.convertToBeamRel(sqlString);
+    String beamPlan = RelOptUtil.toString(exeTree);
+    LOG.info(String.format("beamPlan>\n%s", beamPlan));
+
+    return beamPlan;
+  }
+
+  /**
+   * compile SQL, and return a {@link Pipeline}.
+   *
+   * <p>This may be changed with following CLI client.
+   */
+  @Experimental
+  public static PCollection<BeamSQLRow> compilePipeline(String sqlStatement) throws Exception{
+    PipelineOptions options = PipelineOptionsFactory.fromArgs(new String[] {}).withValidation()
+        .as(PipelineOptions.class); // FlinkPipelineOptions.class
+    options.setJobName("BeamPlanCreator");
+    Pipeline pipeline = Pipeline.create(options);
+
+    return compilePipeline(sqlStatement, pipeline);
+  }
+
+  /**
+   * compile SQL, and return a {@link Pipeline}.
+   *
+   * <p>This may be changed with following CLI client.
+   */
+  @Experimental
+  public static PCollection<BeamSQLRow> compilePipeline(String sqlStatement, Pipeline basePipeline)
+      throws Exception{
+    PCollection<BeamSQLRow> resultStream = planner.compileBeamPipeline(sqlStatement, basePipeline);
+    return resultStream;
+  }
+
+  //methods for Beam SQL as DSL.
+  /**
+   * Translate a PCollection with type T to {@code PCollection<BeamSQLRow>}
+   * which can be accepted by Beam SQL DSL.
+   */
+  public static <T> PTransform<PCollection<T>, PCollection<BeamSQLRow>> toBeamSqlRow(
+      PTransform<PCollection<T>, PCollection<BeamSQLRow>> recordTransform,
```

Comment:
Agree that `toBeamSqlRow` is not must, it's more straight-forward to let users do it in their way.

I don't plan to include the utilities like `BeamSqlRows.fromCsv` here, and would leave it for further discussion as that's more about which default format is supported IMO.